### PR TITLE
Add upgrade agents global action

### PR DESCRIPTION
**Pull request by lucianogorza**
### Description
Add new functionality to the agent table so the user can select agents and trigger a bulk upgrade.
 
### Issues Resolved
https://github.com/wazuh/wazuh-dashboard-plugins/issues/6496

### Evidence
image
image
image
image
image
image
image
image

Command to create outdated agent with docker:
```
docker run --name os-dev-2120-agent-$(date +%s) --network os-dev-2.12.0 --label com.docker.compose.project=os-dev-2120 --env WAZUH_AGENT_VERSION=4.6.0 -d ubuntu:20.04 bash -c '
  apt update -y
  apt install -y curl lsb-release
  curl -so \wazuh-agent-${WAZUH_AGENT_VERSION}.deb \
    https://packages.wazuh.com/4.x/apt/pool/main/w/wazuh-agent/wazuh-agent_${WAZUH_AGENT_VERSION}-1_amd64.deb \
    && WAZUH_MANAGER='wazuh.manager' WAZUH_AGENT_GROUP='default' dpkg -i ./wazuh-agent-${WAZUH_AGENT_VERSION}.deb

  /etc/init.d/wazuh-agent start
  tail -f /var/ossec/logs/ossec.log
'
```

Legend:
:black_circle:: none
:green_circle:: pass
:yellow_circle:: warning
:red_circle:: fail
:white_circle:: not applicable

### General

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Click `More` button: It must show global actions: Add groups to agents, Remove groups from agents, Upgrade agents, Upgrade task details | :black_circle: | :black_circle: | :black_circle: |
| Click `More` button and select the ` Upgrade task details` option. A modal with upgrade tasks must show | :black_circle: | :black_circle: | :black_circle: |
| Without select agents, the global action `Upgrade agents` must be disabled | :black_circle: | :black_circle: | :black_circle: |
| Select agents, the global action `Upgrade agents` must be enabled | :black_circle: | :black_circle: | :black_circle: |
| With read-only user: Select agents, the global action `Upgrade agents` must be disabled | :black_circle: | :black_circle: | :black_circle: |

**Details**
<details>
<summary>:black_circle: Click `More` button: It must show global actions: Add groups to agents, Remove groups from agents, Upgrade agents, Upgrade task details</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Click `More` button and select the ` Upgrade task details` option. A modal with upgrade tasks must show</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Without select agents, the global action `Upgrade agents` must be disabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Select agents, the global action `Upgrade agents` must be enabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: With read-only user: Select agents, the global action `Upgrade agents` must be disabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

### Upgrade agents

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Select two online and outdated agents and a disconnected agent, and click `More` -> `Upgrade agents`. Stop one outdated agent and confirm the upgrade. It must create two upgrade tasks and show one error | :black_circle: | :black_circle: | :black_circle: |
| Check the progress alert (API requests must be triggered every 3 seconds) and wait until the upgrade is finished (API requests must be stopped) | :black_circle: | :black_circle: | :black_circle: |
| Check the final alters, it has to be one failed and one success upgrade task | :black_circle: | :black_circle: | :black_circle: |
| On the upgrade tasks modal, check different filters | :black_circle: | :black_circle: | :black_circle: |

**Details**

<details>
<summary>:black_circle: Select two online and outdated agents and a disconnected agent, and click `More` -> `Upgrade agents`. Stop one outdated agent and confirm the upgrade. It must create two upgrade tasks and show one error</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Check the progress alert (API requests must be triggered every 3 seconds) and wait until the upgrade is finished (API requests must be stopped)</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Check the final alters, it has to be one failed and one success upgrade task</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: On the upgrade tasks modal, check different filters</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: plugins/main/public/components/endpoints-summary/table/upgradeTaskDetailsModal.tsx
Comment (yenienserrano) [inline plugins/main/public/components/endpoints-summary/table/upgradeTaskDetailsModal.tsx:1]: perhaps the file name should be kebab-case like the other files

Comment (lucianogorza) [reply plugins/main/public/components/endpoints-summary/table/upgradeTaskDetailsModal.tsx:1]: Fixed

**Comment by yenienserrano** [comment]
Legend:
:black_circle:: none
:green_circle:: pass
:yellow_circle:: warning
:red_circle:: fail
:white_circle:: not applicable

## UI

### General

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Click `More` button: It must show global actions: Add groups to agents, Remove groups from agents, Upgrade agents, Upgrade task details | :black_circle: | :black_circle: | :black_circle: |
| Click `More` button and select the ` Upgrade task details` option. A modal with upgrade tasks must show | :black_circle: | :black_circle: | :black_circle: |
| Without select agents, the global action `Upgrade agents` must be disabled | :black_circle: | :black_circle: | :black_circle: |
| Select agents, the global action `Upgrade agents` must be enabled | :black_circle: | :black_circle: | :black_circle: |
| With read-only user: Select agents, the global action `Upgrade agents` must be disabled | :black_circle: | :black_circle: | :black_circle: |

**Details**
<details>
<summary>:black_circle: Click `More` button: It must show global actions: Add groups to agents, Remove groups from agents, Upgrade agents, Upgrade task details</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Click `More` button and select the ` Upgrade task details` option. A modal with upgrade tasks must show</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Without select agents, the global action `Upgrade agents` must be disabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Select agents, the global action `Upgrade agents` must be enabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: With read-only user: Select agents, the global action `Upgrade agents` must be disabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

### Upgrade agents

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Select two online and outdated agents and a disconnected agent, and click `More` -> `Upgrade agents`. Stop one outdated agent and confirm the upgrade. It must create two upgrade tasks and show one error | :black_circle: | :black_circle: | :black_circle: |
| Check the progress alert (API requests must be triggered every 3 seconds) and wait until the upgrade is finished (API requests must be stopped) | :black_circle: | :black_circle: | :black_circle: |
| Check the final alters, it has to be one failed and one success upgrade task | :black_circle: | :black_circle: | :black_circle: |
| On the upgrade tasks modal, check different filters | :black_circle: | :black_circle: | :black_circle: |

**Details**

<details>
<summary>:black_circle: Select two online and outdated agents and a disconnected agent, and click `More` -> `Upgrade agents`. Stop one outdated agent and confirm the upgrade. It must create two upgrade tasks and show one error</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Check the progress alert (API requests must be triggered every 3 seconds) and wait until the upgrade is finished (API requests must be stopped)</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Check the final alters, it has to be one failed and one success upgrade task</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: On the upgrade tasks modal, check different filters</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

**Review by yenienserrano — APPROVED** [review]
LGTM

**Comment by asteriscos** [comment]
## UI

### Suggestions

- This field says its not a searchable field, but in the search bar, there's clearly a query using it. This can be confusing.
image

- Whenever something is disable it would be useful to know why. This can be improved by adding a tooltip with the reason of it being disabled or instructions to perform that action.
Screenshot from 2024-03-19 12-47-43


### General

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Click `More` button: It must show global actions: Add groups to agents, Remove groups from agents, Upgrade agents, Upgrade task details | :green_circle: | :black_circle: | :black_circle: |
| Click `More` button and select the ` Upgrade task details` option. A modal with upgrade tasks must show | :green_circle: | :black_circle: | :black_circle: |
| Without select agents, the global action `Upgrade agents` must be disabled | :black_circle: | :black_circle: | :black_circle: |
| Select agents, the global action `Upgrade agents` must be enabled | :black_circle: | :black_circle: | :black_circle: |
| With read-only user: Select agents, the global action `Upgrade agents` must be disabled | :black_circle: | :black_circle: | :black_circle: |

**Details**
<details>
<summary>:green_circle: Click `More` button: It must show global actions: Add groups to agents, Remove groups from agents, Upgrade agents, Upgrade task details</summary>

Chrome - :green_circle:

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Click `More` button and select the ` Upgrade task details` option. A modal with upgrade tasks must show</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Without select agents, the global action `Upgrade agents` must be disabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Select agents, the global action `Upgrade agents` must be enabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: With read-only user: Select agents, the global action `Upgrade agents` must be disabled</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

### Upgrade agents

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Select two online and outdated agents and a disconnected agent, and click `More` -> `Upgrade agents`. Stop one outdated agent and confirm the upgrade. It must create two upgrade tasks and show one error | :black_circle: | :black_circle: | :black_circle: |
| Check the progress alert (API requests must be triggered every 3 seconds) and wait until the upgrade is finished (API requests must be stopped) | :black_circle: | :black_circle: | :black_circle: |
| Check the final alters, it has to be one failed and one success upgrade task | :black_circle: | :black_circle: | :black_circle: |
| On the upgrade tasks modal, check different filters | :black_circle: | :black_circle: | :black_circle: |

**Details**

<details>
<summary>:black_circle: Select two online and outdated agents and a disconnected agent, and click `More` -> `Upgrade agents`. Stop one outdated agent and confirm the upgrade. It must create two upgrade tasks and show one error</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Check the progress alert (API requests must be triggered every 3 seconds) and wait until the upgrade is finished (API requests must be stopped)</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Check the final alters, it has to be one failed and one success upgrade task</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: On the upgrade tasks modal, check different filters</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

**Review by asteriscos — APPROVED** [review]

**Pull request merged**